### PR TITLE
fix: filter emitted events [WPB-18770]

### PR DESCRIPTION
**Pull request by Garzas**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-18770" title="WPB-18770" target="_blank">TaskWPB-18770</a>  [Android] sync incorrectly switching between live and pending when processing a large number of events
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With @vitorhugods help:

- Fixed incorrect filtering of already emitted unprocessed events. Previously, the code used `subList(emittedIndex + 1, eventEntities.lastIndex)`, which mistakenly omitted the last event. Now it correctly uses `.size` to include all relevant items.
- Added dedicated regression tests in `EventRepositoryTest` to confirm that:
  - previously emitted events are not re-emitted,
  - edge cases (e.g., last processed being second-to-last) are handled gracefully.
- Refactored `EventGatherer.gatherEvents()`
  - Replaced previous `channelFlow + Job` logic with `flatMapConcat` on a single `Flow<Unit>`.
  - This ensures proper *sequential* collection:
    - The next batch is not emitted until all events from the current batch are fully processed.
    - Eliminates race conditions between database observation and event processing logic.

### CODE REVIEW

**Review by vitorhugods — APPROVED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/incremental/EventGatherer.kt [hunk 1]
```diff
@@ -85,25 +84,22 @@ internal class EventGathererImpl(
...
-                    } else {
-                        send(EventStreamData.NewEvents(events))
-                    }
-                }
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override suspend fun gatherEvents(): Flow<EventStreamData> {
+        var hasEmitted = false
+
+        // TODO(refactor): stop emitting Unit, emit status of the underlying event fetching for clarity
+        return receiveEventsFromRemoteAndInsertIntoLocalStorage().transform {
```
Comment (sbakhtiarov) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/incremental/EventGatherer.kt:92]: Could this be replaced with first() ?

**Review by MohamadJaara — APPROVED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/incremental/EventGatherer.kt (hunk 1 above)
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/incremental/EventGatherer.kt:92]: We want to avoid `channelFlow`, as it will detach flow emission from collection.

With `channelFlow`, the emitter will not wait for the sender to consume all events before emitting new ones.
